testing/_py/test_local.py

```diff
@@ -207,7 +207,7 @@ def test_visit_norecurse(self, path1):
 
     @pytest.mark.parametrize(
         "fil",
-        ["*dir", "*dir", pytest.mark.skip("sys.version_info <" " (3,6)")(b"*dir")],
+        ["*dir", "*dir", pytest.mark.skip("sys.version_info < (3,6)")(b"*dir")],
```

Comment:
We can remove this:

```suggestion
        ["*dir", "*dir", b"*dir"],
```

Reply:
Right ! Removed the reminder for the whole codebase in [7ded8d8](https://github.com/pytest-dev/pytest/pull/12172/commits/7ded8d8962524787d77dfe30798d777ca7b6685c)

Reply:
yikes, we had skip instead of skipif and the tests where broken

Reply:
965e942dfb4f68d1e28d6388544ff81247a0a4ef introduced the issue replacing a pylib basestring with just str

Reply:
Thank you for looking into it, I was stuck. I'm not sure I understand though, should I replace ``["*dir", "*dir", b"*dir"],`` back to ``["*dir", "*dir", pytest.mark.skipif("sys.version_info < (3,6)")(b"*dir")],`` and then fix the tests ? Maybe in another merge request at this point ?

Reply:
in the Visitor class of the local path clone, there is a condition we need to replace
https://github.com/pytest-dev/pytest/blob/main/src/_pytest/_py/path.py#L140 - it should also check for bytes

Reply:
when asottile took out a lot of cruft a basestring usage was misinterpreted and replaced with just str

unfortunately the skips where wrong and thus the error was never noted as the tests skipped instead of passing

Reply:
I'm not sure that's where things went wrong.

I checked the `py` code, and in Python 3.0+ `py.basestring` was indeed set equal to `str`. The problem is that using `str` and `bytes` interchangeably isn't valid in Python 3, and `py.path.local` never fully accounted for that.

Even in the upstream, this happens in Python 3.6 (or higher, I'm sure):
```python
>>> import py
>>> dir = py.path.local(b'/tmp')
>>> dir
local(b'/tmp')
>>> dir.strpath
b'/tmp'
>>> dir.parts()
[local(b'/'), local(b'/tmp')]
>>> dir.basename
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../venv3.6/lib64/python3.6/site-packages/py/_path/common.py", line 141, in basename
    return self._getbyspec('basename')[0]
  File ".../venv3.6/lib64/python3.6/site-packages/py/_path/local.py", line 293, in _getbyspec
    parts = self.strpath.split(self.sep)
TypeError: a bytes-like object is required, not 'str'
>>> 
>>> # Part of the problem being...
>>> dir.sep
'/'
>>> dir.bestrelpath(b'/')
"b'/'"
>>> # ^^^ Note that that's a _string_ containing "b'/'",
>>> # IOW the result of calling str() on a bytes object
```

...That the tests were being unconditionally skipped, and never caught that, is the bigger problem. But fixing bytes support will _not_ be easy as `_pytest/_py/local.py` (like `py.path.local` before it) is _very_ sloppy with how it handles inputs. When passed `bytes`, things will _sorta_ work... until they very suddenly **really don't**.

Reply:
Oh, and just for fun, I found two _more_ instances of tests being unconditionally skipped by `pytest.mark.skip()`, instead of conditionally with `.skipif()`. Just in that one test file.

https://github.com/pytest-dev/pytest/blob/cf5369db0f14df40c6145b0a90a85577782a13c3/testing/_py/test_local.py#L466-L476

Reply:
thanks for the report; I've opened a new issue to warn when that kind of thing happens 🙏